File: backend/app/services/lead_service.py
from typing import Set, Tuple
from app.models.organization import Organization
from app.schemas.lead import LeadListItem


def calculate_confidence(org: Organization) -> Tuple[float, str]:
    """Calculate confidence score for an organization."""
    score = 0.0

    # Website: 20 points
    if org.websites and any(w.status == "ACTIVE" for w in org.websites):
        score += 20

    # Phone: 20 points
    if org.phone_numbers and len(org.phone_numbers) > 0:
        score += 20

    # Email: 20 points
    if org.email_addresses and len(org.email_addresses) > 0:
        score += 20

    # Address: 20 points
    if org.address or org.city:
        score += 20

    # Contact Person: 10 points
    if org.contacts and len(org.contacts) > 0:
        score += 10

    # Social Links: 10 points
    if org.social_links and len(org.social_links) > 0:
        score += 10

    # Determine level
    if score >= 80:
        level = "HIGH"
    elif score >= 50:
        level = "MEDIUM"
    else:
        level = "LOW"

    return score, level


def build_lead_list_item(org: Organization, saved_ids: Set[int] = None) -> LeadListItem:
    """Build a flat LeadListItem from an Organization ORM object."""
    main_phone = None
    main_whatsapp = None
    if org.phone_numbers:
        for p in org.phone_numbers:
            if p.is_whatsapp and not main_whatsapp:
                main_whatsapp = p.normalized or p.number
            if not p.is_whatsapp and not main_phone:
                main_phone = p.normalized or p.number

    main_email = None
    if org.email_addresses:
        main_email = org.email_addresses[0].email

    main_website = None
    if org.websites:
        for w in org.websites:
            if w.is_official:
                main_website = w.url
                break
        if not main_website:
            main_website = org.websites[0].url

    contact_person = None
    if org.contacts:
        contact_person = org.contacts[0].name

    return LeadListItem(
        id=org.id,
        name=org.name,
        category=org.category,
        location=org.location,
        city=org.city,
        state=org.state,
        confidence_score=org.confidence_score or 0.0,
        confidence_level=org.confidence_level or "LOW",
        is_saved=(saved_ids is not None and org.id in saved_ids),
        created_at=org.created_at,
        main_phone=main_phone,
        main_email=main_email,
        main_website=main_website,
        main_whatsapp=main_whatsapp,
        contact_person=contact_person,
    )

File: backend/src/services/leadService.js
function calculateConfidence(org) {
  let score = 0.0;

  // Website: 20 points
  if (org.websites && org.websites.some(w => w.status === 'ACTIVE' || !w.status)) {
    score += 20;
  }

  // Phone: 20 points
  if (org.phone_numbers && org.phone_numbers.length > 0) {
    score += 20;
  }

  // Email: 20 points
  if (org.email_addresses && org.email_addresses.length > 0) {
    score += 20;
  }

  // Address: 20 points
  if (org.address || org.city) {
    score += 20;
  }

  // Contact Person: 10 points
  if (org.contacts && org.contacts.length > 0) {
    score += 10;
  }

  // Social Links: 10 points
  if (org.social_links && org.social_links.length > 0) {
    score += 10;
  }

  // Determine level
  let level = 'LOW';
  if (score >= 80) {
    level = 'HIGH';
  } else if (score >= 50) {
    level = 'MEDIUM';
  }

  return { score, level };
}

function buildLeadListItem(org, savedIds = new Set()) {
  const plainOrg = org.toJSON ? org.toJSON() : org;

  let main_phone = null;
  let main_whatsapp = null;
  if (plainOrg.phone_numbers && plainOrg.phone_numbers.length > 0) {
    for (const p of plainOrg.phone_numbers) {
      if (p.is_whatsapp && !main_whatsapp) {
        main_whatsapp = p.normalized || p.number;
      }
      if (!p.is_whatsapp && !main_phone) {
        main_phone = p.normalized || p.number;
      }
    }
  }

  let main_email = null;
  if (plainOrg.email_addresses && plainOrg.email_addresses.length > 0) {
    main_email = plainOrg.email_addresses[0].email;
  }

  let main_website = null;
  if (plainOrg.websites && plainOrg.websites.length > 0) {
    for (const w of plainOrg.websites) {
      if (w.is_official) {
        main_website = w.url;
        break;
      }
    }
    if (!main_website) {
      main_website = plainOrg.websites[0].url;
    }
  }

  let contact_person = null;
  if (plainOrg.contacts && plainOrg.contacts.length > 0) {
    contact_person = plainOrg.contacts[0].name;
  }

  const isSaved = savedIds instanceof Set ? savedIds.has(plainOrg.id) : (Array.isArray(savedIds) ? savedIds.includes(plainOrg.id) : false);

  return {
    id: plainOrg.id,
    name: plainOrg.name,
    category: plainOrg.category,
    location: plainOrg.location,
    city: plainOrg.city,
    state: plainOrg.state,
    confidence_score: plainOrg.confidence_score || 0.0,
    confidence_level: plainOrg.confidence_level || 'LOW',
    is_saved: isSaved,
    created_at: plainOrg.created_at,
    main_phone,
    main_email,
    main_website,
    main_whatsapp,
    contact_person,
  };
}

module.exports = {
  calculateConfidence,
  buildLeadListItem,
};
